src/cpp/server/server_cc.cc

```diff
@@ -934,6 +1006,13 @@ void Server::ShutdownInternal(gpr_timespec deadline) {
       (*it)->Wait();
     }
 
+    // Wait for all outstanding callback requests to complete
+    {
+      std::unique_lock<std::mutex> cblock(callback_reqs_mu_);
+      callback_reqs_done_cv_.wait(
+          cblock, [this] { return callback_reqs_outstanding_ == 0; });
```

Comment:
What prevents new callback_reqs from being created right after this? It doesn't look like we check for shutdown_ before we do that?

Reply:
So this is a measure of total callback requests outstanding, not just those waiting for a match. So you see that callback requests only get created in two places:

1. At server startup time (which won't be shutdown time)
1. When we have a successful match on a request and there are no more requests waiting to match for that method

The second one is the interesting one here. If the server is shutting down, we'll never again have a successful match: we'll report an unsuccessful match for any not-yet-matched requests and delete the "matched" request, decrementing the count. Alternatively, we could also enter a situation where a server successfully matched a request but then decided to shutdown between the time we allocated a new request and tried to register it. If that happens, the register call will report a failure and we'll know that registers are failing because of shutdown. So again we'll never increment the counter again and will decrement it by deleting the newly created request.

Reply:
That should probably be a comment in the code.

Reply:
When the server goes into Shutdown, it cancels all calls that have not yet been matched, so those will all get their CallbackCallTag::Run invoked with ok=false indicating a shutdown, which then causes us to delete the request.